Run "npm ci" before checking npm licenses

This is related to #373.
When running PK on a fresh checkout (without `node_modules`) it fails with the following exception:

```
2022-11-07 08:16:31.374 [FINEST ] Executing command npm list 
2022-11-07 08:16:31.375 [FINEST ] Executing command 'npm list --location=project --depth=0 --json' in working dir '/Users/<user>/git/extension-manager/registry' 
Exception in thread "main" java.lang.IllegalStateException: E-PK-CORE-126: Failed to run command 'npm list --location=project --depth=0 --json' in '/Users/<user>/git/extension-manager/registry', exit code was 1 after PT1.099924S. Output:
'{
  "version": "0.1.0",
  "name": "registry",
  "problems": [
    "missing: @types/jest@^29.2.1, required by registry@0.1.0",
    "missing: @types/node@18.11.9, required by registry@0.1.0",
    "missing: @types/prettier@2.7.1, required by registry@0.1.0",
    "missing: aws-cdk-lib@2.50.0, required by registry@0.1.0",
    "missing: aws-cdk@2.50.0, required by registry@0.1.0",
    "missing: constructs@^10.1.148, required by registry@0.1.0",
    "missing: jest@^29.2.2, required by registry@0.1.0",
    "missing: source-map-support@^0.5.21, required by registry@0.1.0",
    "missing: ts-jest@^29.0.3, required by registry@0.1.0",
    "missing: ts-node@^10.9.1, required by registry@0.1.0",
    "missing: typescript@~4.8.4, required by registry@0.1.0"
  ],
  "dependencies": {
    "@types/jest": {
      "required": "^29.2.1",
      "missing": true,
      "problems": [
        "missing: @types/jest@^29.2.1, required by registry@0.1.0"
      ]
    },
    "@types/node": {
      "missing": true,
      "problems": [
        "missing: @types/node@18.11.9, required by registry@0.1.0"
      ]
    },
    "@types/prettier": {
      "missing": true,
      "problems": [
        "missing: @types/prettier@2.7.1, required by registry@0.1.0"
      ]
    },
    "aws-cdk-lib": {
      "missing": true,
      "problems": [
        "missing: aws-cdk-lib@2.50.0, required by registry@0.1.0"
      ]
    },
    "aws-cdk": {
      "missing": true,
      "problems": [
        "missing: aws-cdk@2.50.0, required by registry@0.1.0"
      ]
    },
    "constructs": {
      "missing": true,
      "problems": [
        "missing: constructs@^10.1.148, required by registry@0.1.0"
      ]
    },
    "jest": {
      "missing": true,
      "problems": [
        "missing: jest@^29.2.2, required by registry@0.1.0"
      ]
    },
    "source-map-support": {
      "missing": true,
      "problems": [
        "missing: source-map-support@^0.5.21, required by registry@0.1.0"
      ]
    },
    "ts-jest": {
      "missing": true,
      "problems": [
        "missing: ts-jest@^29.0.3, required by registry@0.1.0"
      ]
    },
    "ts-node": {
      "missing": true,
      "problems": [
        "missing: ts-node@^10.9.1, required by registry@0.1.0"
      ]
    },
    "typescript": {
      "missing": true,
      "problems": [
        "missing: typescript@~4.8.4, required by registry@0.1.0"
      ]
    }
  }
}
'
Error output:
'npm ERR! code ELSPROBLEMS
npm ERR! missing: @types/jest@^29.2.1, required by registry@0.1.0
npm ERR! missing: @types/node@18.11.9, required by registry@0.1.0
npm ERR! missing: @types/prettier@2.7.1, required by registry@0.1.0
npm ERR! missing: aws-cdk-lib@2.50.0, required by registry@0.1.0
npm ERR! missing: aws-cdk@2.50.0, required by registry@0.1.0
npm ERR! missing: constructs@^10.1.148, required by registry@0.1.0
npm ERR! missing: jest@^29.2.2, required by registry@0.1.0
npm ERR! missing: source-map-support@^0.5.21, required by registry@0.1.0
npm ERR! missing: ts-jest@^29.0.3, required by registry@0.1.0
npm ERR! missing: ts-node@^10.9.1, required by registry@0.1.0
npm ERR! missing: typescript@~4.8.4, required by registry@0.1.0
{
  "error": {
    "code": "ELSPROBLEMS",
    "summary": "missing: @types/jest@^29.2.1, required by registry@0.1.0\nmissing: @types/node@18.11.9, required by registry@0.1.0\nmissing: @types/prettier@2.7.1, required by registry@0.1.0\nmissing: aws-cdk-lib@2.50.0, required by registry@0.1.0\nmissing: aws-cdk@2.50.0, required by registry@0.1.0\nmissing: constructs@^10.1.148, required by registry@0.1.0\nmissing: jest@^29.2.2, required by registry@0.1.0\nmissing: source-map-support@^0.5.21, required by registry@0.1.0\nmissing: ts-jest@^29.0.3, required by registry@0.1.0\nmissing: ts-node@^10.9.1, required by registry@0.1.0\nmissing: typescript@~4.8.4, required by registry@0.1.0",
    "detail": ""
  }
}

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/<user>/.npm/_logs/2022-11-07T07_16_32_320Z-debug-0.log
'
        at com.exasol.projectkeeper.sources.analyze.golang.SimpleProcess.waitUntilFinished(SimpleProcess.java:99)
        at com.exasol.projectkeeper.sources.analyze.generic.CommandExecutor.execute(CommandExecutor.java:25)
        at com.exasol.projectkeeper.sources.analyze.npm.NpmServices.getJsonOutput(NpmServices.java:67)
        at com.exasol.projectkeeper.sources.analyze.npm.NpmServices.listDependencies(NpmServices.java:59)
        at com.exasol.projectkeeper.sources.analyze.npm.NpmDependencies.getDependencies(NpmDependencies.java:35)
        at com.exasol.projectkeeper.sources.analyze.npm.NpmServices.getDependencies(NpmServices.java:46)
        at com.exasol.projectkeeper.sources.analyze.npm.NpmSourceAnalyzer.analyzeSource(NpmSourceAnalyzer.java:52)
        at com.exasol.projectkeeper.sources.analyze.npm.NpmSourceAnalyzer.lambda$analyze$0(NpmSourceAnalyzer.java:36)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at com.exasol.projectkeeper.sources.analyze.npm.NpmSourceAnalyzer.analyze(NpmSourceAnalyzer.java:36)
        at com.exasol.projectkeeper.sources.SourceAnalyzer.analyzeSources(SourceAnalyzer.java:72)
        at com.exasol.projectkeeper.sources.SourceAnalyzer.analyze(SourceAnalyzer.java:61)
        at com.exasol.projectkeeper.ProjectKeeper.phase2(ProjectKeeper.java:126)
        at com.exasol.projectkeeper.ProjectKeeper.runValidationPhases(ProjectKeeper.java:225)
        at com.exasol.projectkeeper.ProjectKeeper.fix(ProjectKeeper.java:211)
        at com.exasol.projectkeeper.cli.ProjectKeeperLauncher.runProjectKeeper(ProjectKeeperLauncher.java:64)
        at com.exasol.projectkeeper.cli.ProjectKeeperLauncher.start(ProjectKeeperLauncher.java:59)
        at com.exasol.projectkeeper.cli.ProjectKeeperLauncher.main(ProjectKeeperLauncher.java:53)
```

## Proposed solution
1. PK runs `npm ci` before running `npm list`.
2. Update `project-keeper-verify.yml`: Install a defined version of node/npm with [actions/setup-node](https://github.com/actions/setup-node) and configure caching
```yaml
- uses: actions/setup-node@v3
  with:
    node-version: 16
    cache: 'npm'
    cache-dependency-path: "**/package-lock.json"
```

--- a/doc/changes/changelog.md
+++ b/doc/changes/changelog.md
@@ -1,5 +1,6 @@
 # Changes
 
+* [2.9.2](changes_2.9.2.md)
 * [2.9.1](changes_2.9.1.md)
 * [2.9.0](changes_2.9.0.md)
 * [2.8.0](changes_2.8.0.md)

--- a/doc/changes/changes_2.9.2.md
+++ b/doc/changes/changes_2.9.2.md
@@ -1,0 +1,55 @@
+# Project Keeper 2.9.2, released 2022-??-??
+
+Code name: Bug Fix
+
+## Summary
+
+Fixed bug in support for NPM projects.
+
+## Bug Fixes
+
+* #403: Added `npm ci` before checking npm licenses
+
+## Dependency Updates
+
+### Project Keeper Core
+
+#### Compile Dependency Updates
+
+* Updated `com.exasol:project-keeper-shared-model-classes:2.9.1` to `2.9.2`
+
+#### Runtime Dependency Updates
+
+* Updated `com.exasol:project-keeper-java-project-crawler:2.9.1` to `2.9.2`
+
+#### Test Dependency Updates
+
+* Updated `com.exasol:project-keeper-shared-test-setup:2.9.1` to `2.9.2`
+
+### Project Keeper Command Line Interface
+
+#### Compile Dependency Updates
+
+* Updated `com.exasol:project-keeper-core:2.9.1` to `2.9.2`
+
+#### Test Dependency Updates
+
+* Updated `com.exasol:project-keeper-shared-test-setup:2.9.1` to `2.9.2`
+
+### Project Keeper Maven Plugin
+
+#### Compile Dependency Updates
+
+* Updated `com.exasol:project-keeper-core:2.9.1` to `2.9.2`
+
+### Project Keeper Java Project Crawler
+
+#### Compile Dependency Updates
+
+* Updated `com.exasol:project-keeper-shared-model-classes:2.9.1` to `2.9.2`
+
+### Project Keeper Shared Test Setup
+
+#### Compile Dependency Updates
+
+* Updated `com.exasol:project-keeper-shared-model-classes:2.9.1` to `2.9.2`

--- a/error_code_config.yml
+++ b/error_code_config.yml
@@ -1,0 +1,6 @@
+# This file is only used for release guide.
+# Actual error codes are based on files error_code_config.yml in individual subfolders.
+
+error-tags:
+  PK:
+    highest-index: 0

--- a/parent-pom/pom.xml
+++ b/parent-pom/pom.xml
@@ -28,7 +28,7 @@
         </repository>
     </distributionManagement>
     <properties>
-        <revision>2.9.1</revision>
+        <revision>2.9.2</revision>
         <maven.version>3.8.6</maven.version>
         <junit.version>5.9.1</junit.version>
         <jacoco.version>0.8.8</jacoco.version>

--- a/pom.xml
+++ b/pom.xml
@@ -82,7 +82,7 @@
                              https://ossindex.sonatype.org/vulnerability/CVE-2021-26291 -->
                         <exclude>CVE-2021-26291</exclude>
                         <!-- Exclude vulnerability Stack-based Buffer Overflow (6.5)
-                             found in transitive dependency org.yaml:snakeyaml:jar:1.31
+                             found in transitive dependency org.yaml:snakeyaml:jar:1.33
                              used by as there is no update available currently
                              This exclude unfortunatly needs to be made in all of the following files:
                              - pom.xml (this file)
@@ -91,7 +91,7 @@
                              - project-keeper-maven-plugin/pom.xml
                              - shared-test-setup/pom.xml
                         -->
-                        <exclude>CVE-2022-38752</exclude>
+                        <exclude>CVE-2022-1471</exclude>
                     </excludeVulnerabilityIds>
                 </configuration>
             </plugin>

--- a/project-keeper-cli/pom.xml
+++ b/project-keeper-cli/pom.xml
@@ -107,7 +107,7 @@
                 <configuration>
                     <excludeVulnerabilityIds>
                         <!-- Affects multiple subprojects, see ../pom.xml. -->
-                        <exclude>CVE-2022-38752</exclude>
+                        <exclude>CVE-2022-1471</exclude>
                     </excludeVulnerabilityIds>
                 </configuration>
             </plugin>

--- a/project-keeper-maven-plugin/pom.xml
+++ b/project-keeper-maven-plugin/pom.xml
@@ -123,7 +123,7 @@
                              https://ossindex.sonatype.org/vulnerability/CVE-2021-26291 -->
                         <exclude>CVE-2021-26291</exclude>
                         <!-- Affects multiple subprojects, see ../pom.xml. -->
-                        <exclude>CVE-2022-38752</exclude>
+                        <exclude>CVE-2022-1471</exclude>
                     </excludeVulnerabilityIds>
                 </configuration>
             </plugin>

--- a/project-keeper/pom.xml
+++ b/project-keeper/pom.xml
@@ -136,7 +136,7 @@
                              https://ossindex.sonatype.org/vulnerability/sonatype-2020-0491 -->
                         <exclude>sonatype-2020-0491</exclude>
                         <!-- Affects multiple subprojects, see ../pom.xml. -->
-                        <exclude>CVE-2022-38752</exclude>
+                        <exclude>CVE-2022-1471</exclude>
                     </excludeVulnerabilityIds>
                 </configuration>
             </plugin>

--- a/project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/npm/NpmServices.java
+++ b/project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/npm/NpmServices.java
@@ -16,6 +16,11 @@ class NpmServices {
     private static final String NPM = "npm" + OsCheck.suffix(".cmd");
     private static final String NPX = "npx" + OsCheck.suffix(".cmd");
 
+    static final ShellCommand FETCH_DEPENDENCIES = ShellCommand.builder() //
+            .command(NPM, "ci") //
+            .timeout(Duration.ofMinutes(5)) //
+            .build();
+
     // [impl -> dsn~npm-dependency-licenses~1]
     static final ShellCommand LIST_DEPENDENCIES = ShellCommand.builder() //
             .command(NPM, "list") //
@@ -32,6 +37,7 @@ class NpmServices {
 
     private final CommandExecutor executor;
     private final GitService git;
+    private boolean fetchedDependencies = false;
 
     NpmServices(final CommandExecutor executor, final GitService git) {
         this.executor = executor;
@@ -64,7 +70,16 @@ class NpmServices {
     }
 
     private JsonObject getJsonOutput(final ShellCommand cmd, final Path workingDir) {
+        fetchDependencies(workingDir);
         final String stdout = this.executor.execute(cmd, workingDir);
         return JsonIo.read(new StringReader(stdout));
+    }
+
+    private void fetchDependencies(final Path folder) {
+        if (this.fetchedDependencies) {
+            return;
+        }
+        this.executor.execute(FETCH_DEPENDENCIES, folder);
+        this.fetchedDependencies = true;
     }
 }

--- a/project-keeper/src/main/resources/non_maven_templates/.github/workflows/project-keeper-verify.yml
+++ b/project-keeper/src/main/resources/non_maven_templates/.github/workflows/project-keeper-verify.yml
@@ -29,6 +29,13 @@ jobs:
         with:
           go-version: 1.19
 
+      - name Set up NPM
+        uses: actions/setup-node@v3
+        with:
+          node-version: 16
+          cache: 'npm'
+          cache-dependency-path: "**/package-lock.json"
+
       - name: Cache Project Keeper
         uses: actions/cache@v3
         with:

--- a/project-keeper/src/test/java/com/exasol/projectkeeper/sources/analyze/npm/NpmServicesTest.java
+++ b/project-keeper/src/test/java/com/exasol/projectkeeper/sources/analyze/npm/NpmServicesTest.java
@@ -1,7 +1,6 @@
 package com.exasol.projectkeeper.sources.analyze.npm;
 
-import static com.exasol.projectkeeper.sources.analyze.npm.NpmServices.LICENSE_CHECKER;
-import static com.exasol.projectkeeper.sources.analyze.npm.NpmServices.LIST_DEPENDENCIES;
+import static com.exasol.projectkeeper.sources.analyze.npm.NpmServices.*;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.*;
 import static org.mockito.ArgumentMatchers.any;
@@ -35,12 +34,12 @@ class NpmServicesTest {
     @Mock
     TaggedCommit previousTag;
 
-    private static final String PREVIOUS_VERSION = "1.1.1";
     private static final Path PROJECT_DIR = Paths.get("/projects/sample-project");
     private static final Path PACKAGE_JSON_FILE = Paths.get("some/file.txt");
 
     @Test
     void getDependencies() {
+        when(this.executor.execute(eq(FETCH_DEPENDENCIES), any())).thenReturn("");
         when(this.executor.execute(eq(LIST_DEPENDENCIES), any())).thenReturn(TestData.DEPENDENCIES);
         when(this.executor.execute(eq(LICENSE_CHECKER), any())).thenReturn(TestData.LICENSES);
         final PackageJson current = TestData.samplePackageJson();

--- a/project-keeper/src/test/java/com/exasol/projectkeeper/validators/dependencies/DependenciesValidatorIT.java
+++ b/project-keeper/src/test/java/com/exasol/projectkeeper/validators/dependencies/DependenciesValidatorIT.java
@@ -104,7 +104,7 @@ class DependenciesValidatorIT extends ProjectKeeperAbstractMavenIT {
         final TestMavenModel pomModel = new TestMavenModel();
         pomModel.configureAssemblyPluginFinalName();
         final Repository exasolRepo = new Repository();
-        exasolRepo.setUrl("https://maven.exasol.com/artifactory/exasol-releases");
+        exasolRepo.setUrl("https://repo1.maven.org/maven2");
         exasolRepo.setId("maven.exasol.com");
         pomModel.addRepository(exasolRepo);
         pomModel.addDependency("exasol-jdbc", "com.exasol", "compile", "7.0.4");

--- a/release_config.yml
+++ b/release_config.yml
@@ -2,3 +2,6 @@ release-platforms:
   - GitHub
   - Maven
 language: Java
+maven-artifacts:
+  - exasol/project-keeper-cli
+  - exasol/project-keeper-maven-plugin

--- a/shared-test-setup/pom.xml
+++ b/shared-test-setup/pom.xml
@@ -42,7 +42,7 @@
                 <configuration>
                     <excludeVulnerabilityIds>
                         <!-- Affects multiple subprojects, see ../pom.xml. -->
-                        <exclude>CVE-2022-38752</exclude>
+                        <exclude>CVE-2022-1471</exclude>
                     </excludeVulnerabilityIds>
                 </configuration>
             </plugin>

